db/db_fingerprint.c

```diff
@@ -0,0 +1,122 @@
+        if (zNormSql != NULL) {
+            *pzNormSql = strdup(zNormSql);
+        }
+    } else if (gbl_verbose_normalized_queries) {
+        logmsg(LOGMSG_ERROR,
+               "FAILED sqlite3_prepare_v2(%p, {%s}) for normalization, rc=%d, msg=%s\n",
+               db, zSql, rc, sqlite3_errmsg(db));
+    }
+    sqlite3_finalize(p); /* p may be NULL and that is OK */
+}
+
+void add_fingerprint(sqlite3 *sqldb, int64_t cost, int64_t time, int64_t nrows,
+                     char *sql) {
+    char *zNormSql = NULL;
+
+    normalize_query(sqldb, sql, &zNormSql);
+    if (zNormSql != NULL) {
+        size_t nNormSql = strlen(zNormSql);
+        unsigned char fingerprint[FINGERPRINTSZ];
+        MD5Context ctx;
+        MD5Init(&ctx);
+        MD5Update(&ctx, (unsigned char *)zNormSql, nNormSql);
+        memset(fingerprint, 0, sizeof(fingerprint));
+        MD5Final(fingerprint, &ctx);
+        Pthread_mutex_lock(&gbl_fingerprint_hash_mu);
+        if (gbl_fingerprint_hash == NULL) gbl_fingerprint_hash = hash_init(FINGERPRINTSZ);
+        struct fingerprint_track *t = hash_find(gbl_fingerprint_hash, fingerprint);
+        if (t == NULL) {
+            /* make sure we haven't generated an unreasonable number of these */
+            int nents;
+            hash_info(gbl_fingerprint_hash, NULL, NULL, NULL, NULL, &nents, NULL, NULL);
+            if (nents >= gbl_fingerprint_max_queries) {
+                static int complain_once = 1;
+                if (complain_once) {
+                    logmsg(LOGMSG_WARN,
+                           "Stopped tracking fingerprints, hit max #queries %d.\n",
+                           gbl_fingerprint_max_queries);
+                    complain_once = 0;
+                }
+                Pthread_mutex_unlock(&gbl_fingerprint_hash_mu);
+                free(zNormSql);
+                goto done;
+            }
+            t = calloc(1, sizeof(struct fingerprint_track));
+            memcpy(t->fingerprint, fingerprint, FINGERPRINTSZ);
+            t->count = 1;
+            t->cost = cost;
+            t->time = time;
+            t->rows = nrows;
+            t->zNormSql = zNormSql;
+            t->nNormSql = nNormSql;
+            hash_add(gbl_fingerprint_hash, t);
+            if (gbl_verbose_normalized_queries) {
+                char fp[FINGERPRINTSZ*2+1]; /* 16 ==> 33 */
+                util_tohex(fp, t->fingerprint, FINGERPRINTSZ);
+                logmsg(LOGMSG_USER, "NORMALIZED [%s] {%s} ==> {%s}\n",
+                       fp, sql, t->zNormSql);
+            }
+        } else {
+            t->count++;
+            t->cost += cost;
+            t->time += time;
+            t->rows += nrows;
+            assert( memcmp(t->fingerprint,fingerprint,FINGERPRINTSZ)==0 );
+            assert( t->zNormSql!=zNormSql );
+            assert( t->nNormSql==nNormSql );
+            assert( strncmp(t->zNormSql,zNormSql,t->nNormSql)==0 );
+            free(zNormSql);
+        }
```

Comment:
You need to call `reqlog_set_fingerprint(logger, fingerprint);` at the end. Otherwise reqlog will not have the fingerprint.